### PR TITLE
bump github.com/coreos/go-oidc dep to pick up nbf enforcement

### PR DESCRIPTION
**Pull request by anderseknert**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
If the nbf (not before) claim is present in an ID token, check that it's indeed in the past. This allows the token issuer to control the point in time at when the token should be usable for authentication.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
The nbf (not before) claim, if present in ID token, is now enforced.  
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 1]
```diff
@@ -633,6 +636,21 @@ func (a *Authenticator) AuthenticateToken(ctx context.Context, token string) (*a
...
 
+	// Ensure not-before (nbf) claim, if present, is indeed in the past.
+	if c.hasClaim("nbf") {
+		var nbfValue int64
+		if err := c.unmarshalClaim("nbf", &nbfValue); err != nil {
+			return nil, false, fmt.Errorf("oidc: parse claim nbf: %v", err)
+		}
+
+		nbfTime := time.Unix(nbfValue, 0)
+		now := a.now()
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: now() can be nil here. This needs to default to time.Now above.

**Comment by ericchiang** [comment]
This might be a better feature request for github.com/coreos/go-oidc, since that's what does the current expiration validation.

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go (hunk 1 above)
Comment (anderseknert) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Ah, alright - golang new to me so still trying to figure out all the conditions for when nil checks are needed. I thought the check at line 308 would be enough, but coming to think of it of course that only covers the case where the authenticator is constructed by the `newAuthenticator` function. Any good tools for finding possible nil-problems like there is for Java? I'll fix.

**Comment by anderseknert** [comment]
I considered that too, but `nbf´ doesn't make much sense outside of authentication contexts IMO - and as that library is more generic I thought it would fit better here. If you think I should try submitting a PR there first I'll do that.

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go (hunk 1 above)
Comment (ericchiang) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: I don't know of a common one. Generally, you ensure this during initialization:

```
now := config.now
if now == nil {
    now = time.Now
}
a := authenticator{
    now: now,
    // ...
}
```

**Comment by ericchiang** [comment]
I think `nbf` makes sense for ID Token validation in go-oidc instead of having it only in Kubernetes. Want to submit it there, then we can pull it into kube?

cc @mikedanese

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go (hunk 1 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Or  `a := authenticator{now: time.Now}` and set over it in tests.

**Review by mikedanese — COMMENTED** [review]
Adding this to go-oidc makes sense to me as well. Can you send a change to https://github.com/coreos/go-oidc, we'll review there, and then we'll bump the dep?

**Comment by anderseknert** [comment]
Thanks guys! PR targeting go-oidc coming up 👍

**Comment by mikedanese** [comment]
@anderseknert do you want to bump the go-oidc dependency?

**Comment by anderseknert** [comment]
I'd be delighted @mikedanese . Just heading to bed here but will check it out first thing tomorrow after having read up on how that works, hah. It's been three weeks since I wrote my first line of Go so it's all been moving pretty fast. Any pointers would be appreciated :)

**Comment by mikedanese** [comment]
updating deps in kubernetes is no easy task. will give it a shot and report instructions.

**Comment by mikedanese** [comment]
/lgtm
/hold

This needs a release note. Can you update the description?

**Comment by anderseknert** [comment]
@mikedanese ✅

File: go.mod [hunk 2]
```diff
@@ -219,7 +219,7 @@ replace (
 	github.com/coreos/bbolt => github.com/coreos/bbolt v1.3.1-coreos.6
 	github.com/coreos/etcd => github.com/coreos/etcd v3.3.13+incompatible
 	github.com/coreos/go-etcd => github.com/coreos/go-etcd v2.0.0+incompatible
-	github.com/coreos/go-oidc => github.com/coreos/go-oidc v0.0.0-20180117170138-065b426bd416
+	github.com/coreos/go-oidc => github.com/coreos/go-oidc v2.0.1-0.20190815175729-2be1c5b8a260+incompatible
```
Comment (liggitt) [inline go.mod]: I won't stick on it since we weren't on a tag before, but any chance of getting a v2.1.0 tag cut for this? cc @ericchiang

Comment (ericchiang) [reply go.mod]: Yeah. I'll do that today

**Comment by liggitt** [comment]
> Edit: short answer is it's really hard. Maybe there are some docs I don't know about.

See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/vendor.md#adding-or-updating-a-dependency

File: go.mod (hunk 2 above)
Comment (ericchiang) [reply go.mod]: Tagged https://github.com/coreos/go-oidc/releases/tag/v2.1.0 :)

**Comment by anderseknert** [comment]
That's a great resource, thanks @liggitt! Will try to do it myself next time. About the failing tests reported, I don't see how they are related to any changes made here. Are they known to be flaky, or something else that changed? I'm on a 2 week leave so a little hard trying to follow when not on my computer. Again - thanks for all the help guys!

**Comment by liggitt** [comment]
unfortunately, yes, sometimes the scale tests flake

**Comment by anderseknert** [comment]
Anything more needed from me here?

**Comment by anderseknert** [comment]
@liggitt I've now rebased, fixed conflicts and run `hack/update-vendor.sh` which checked out (thanks for the pointer, btw). As the conflicting changes are in go.mod|sum though I expect this will diverge from master soon again though.

**Comment by liggitt** [comment]
Thanks!

/lgtm

**Pull request merged**